src/libstd/rt/unwind/seh64_gnu.rs

```diff
@@ -0,0 +1,231 @@
+    let params = [Box::into_raw(panic_ctx) as ULONG_PTR];
+    rtdebug!("panic: ctx={:X}", params[0]);
+    RaiseException(RUST_PANIC,
+                   EXCEPTION_NONCONTINUABLE,
+                   params.len() as DWORD,
+                   &params as *const ULONG_PTR);
+    rtabort!("could not unwind stack");
+}
+
+pub unsafe fn cleanup(ptr: *mut c_void) -> Box<Any + Send + 'static> {
+    rtdebug!("cleanup: ctx={:X}", ptr as usize);
+    let panic_ctx = Box::from_raw(ptr as *mut PanicData);
+    return panic_ctx.data;
+}
+
+// SEH doesn't support resuming unwinds after calling a landing pad like libunwind does.
+// For this reason, MSVC compiler outlines landing pads into separate functions that
+// can be called directly from the personality function but are nevertheless able to find
+// and modify stack frame of the "parent" function.
+//
+// Since this cannot be done with libdwarf-style landing pads, rust_eh_personality instead
+// catches RUST_PANICs, runs the landing pad, then reraises the exception.
+//
+// Note that it makes certain assumptions about the exception:
+// 1. That RUST_PANIC is non-continuable, so no lower stack frame may choose to resume execution.
+// 2. That the first parameter of the exception is a pointer to an extra data area (PanicData).
+// Since these assumptions do not generally hold true for foreign exceptions (system faults,
+// C++ exceptions, etc), we make no attempt to invoke our landing pads (and, thus, destructors!)
+// for anything other than RUST_PANICs. This is considered acceptable, because the behavior of
+// throwing exceptions through a C ABI boundary is undefined.
+
+#[no_mangle] // referenced from rust_try.ll
+#[cfg(not(test))]
+pub unsafe extern "C" fn rust_eh_personality_catch(
+    exceptionRecord: *mut EXCEPTION_RECORD,
+    establisherFrame: LPVOID,
+    contextRecord: *mut CONTEXT,
+    dispatcherContext: *mut DISPATCHER_CONTEXT
+) -> EXCEPTION_DISPOSITION
+{
+    let er = &*exceptionRecord;
+    let dc = &*dispatcherContext;
+    rtdebug!("rust_eh_personality_catch: code={:X}, flags={:X}, frame={:X}, ip={:X}",
+        er.ExceptionCode, er.ExceptionFlags, establisherFrame as usize, dc.ControlPc as usize);
+
+    if er.ExceptionFlags & EXCEPTION_UNWIND == 0 { // we are in the dispatch phase
+        if er.ExceptionCode == RUST_PANIC {
+            if let Some(lpad) = find_landing_pad(dc) {
+                RtlUnwindEx(establisherFrame,
+                            lpad as LPVOID,
+                            exceptionRecord,
+                            er.ExceptionInformation[0] as LPVOID, // pointer to PanicData
+                            contextRecord,
+                            dc.HistoryTable);
+                rtabort!("could not unwind");
+            }
+            rtabort!("could not locate the landing pad");
+        }
+    }
+    ExceptionContinueSearch
+}
+
+#[lang="eh_personality"]
+#[cfg(not(test))]
+unsafe extern "C" fn rust_eh_personality(
```

Comment:
Is this just the same as `rust_eh_personality_catch`? These two functions look the same except for the expectation of the landing pad existing...

Overall I'd find it at least helpful to have a nice big doc comment at the top of the module explaining what's going on here. I have a little understanding of what mechanisms are in play, but I know why these are structured the way they are. For example it's not clear to my why the catch personality is catching but this one isn't?